rest-service/manager_rest/cluster_status_manager.py

```diff
@@ -0,0 +1,237 @@
+from manager_rest.rest.rest_utils import parse_datetime_string
+
+
+DB_SERVICE_KEY = 'PostgreSQL'
+BROKER_SERVICE_KEY = 'RabbitMQ'
+UNINITIALIZED_STATUS = 'Uninitialized'
+CLUSTER_STATUS_PATH = '/opt/manager/resources/cluster_status'
+
+
+class ClusterStatusManager(object):
+    def __init__(self):
+        self.cloudify_version = ''
+        self.cluster_status = {}
+        self.cluster_structure = {}
+        self.missing_status_files = {}
+
+    def get_cluster_status(self):
+        """
+        Generate the cluster status using:
+        1. The DB tables (managers, rabbitmq_brokers, db_nodes) for the
+           structure of the cluster.
+        2. The status files each reporter sends with the most updated status
+           of the node.
+        """
+        self.cluster_status = {'status': ServiceStatus.DEGRADED,
+                               'services': {}}
+
+        if not path.isdir(CLUSTER_STATUS_PATH):
+            return self.cluster_status
+
+        self.cluster_structure = self._get_cluster_structure()
+        self.cloudify_version = \
+            self.cluster_structure[ServiceType.MANAGER][0].version
+        self.missing_status_files = {}
+
+        for service_type, service_nodes in self.cluster_structure.items():
+            self._generate_service_nodes_status(service_type, service_nodes)
+
+        self._handle_missing_status_files()
+        self.cluster_status['status'] = self._get_entire_cluster_status()
+        return self.cluster_status
+
+    def _get_cluster_structure(self):
+        storage_manager = get_storage_manager()
+        return {
+            ServiceType.MANAGER: storage_manager.list(models.Manager),
+            ServiceType.BROKER: storage_manager.list(models.RabbitMQBroker),
+            ServiceType.DB: storage_manager.list(models.DBNodes)
+        }
+
+    def _generate_service_nodes_status(self, service_type, service_nodes):
+        formatted_nodes = {}
+
+        for node in service_nodes:
+            file_name = '{0}_{1}.json'.format(service_type, node.node_id)
+            status_file_path = path.join(CLUSTER_STATUS_PATH, file_name)
+
+            if not path.exists(status_file_path):
+                self._add_missing_status_file(node, formatted_nodes,
+                                              service_type)
+                continue
+
+            with open(status_file_path, 'r') as status_file:
+                node_status = json.load(status_file)
+
+            if not self._is_report_valid(node_status):
+                self._generate_node_status(node, formatted_nodes,
+                                           status=ServiceStatus.FAIL)
+                continue
+
+            report = node_status['report']
+            self._generate_node_status(node, formatted_nodes,
+                                       report['services'], report['status'])
+
+        is_external = service_nodes[0].to_dict().get('is_external', False)
```

Comment:
no real reason for the `.to_dict`? vs `service_nodes[0].is_external`

Reply:
same for other usages of `.to_dict` in the file

Reply:
It is for cases it doesn't have this column, for example the managers table in the DB doesn't have the `is_external` column and the rabbitmq_brokers and db_nodes don't have the `public_ip` column.

Reply:
I added the `CloudifyNodeMixin` and now I don't need the `to_dict`